// reconnect.js/build/__TypeCheck__.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var index_1 = require("./index");
// this function is absolutely NON-SENSE
// because it's only used to verify the type definition in IDE
function useTypeCheck() {
    index_1.useNewOutlet('my-outlet', function () { return 100; });
    var _a = index_1.useOutlet('my-outlet'), value = _a[0], setValue = _a[1];
    var udpateInt = index_1.useOutletSetter('my-outlet');
    function testSetter() {
        setValue(value + 1);
    }
    function testSetterWithCallback() {
        setValue(function (oldValue) {
            return oldValue + 1;
        });
    }
    function testUpdateInt() {
        udpateInt(100);
    }
    function testUpdateIntCb() {
        udpateInt(function (value) {
            return value + 3;
        });
    }
    return {
        testSetter: testSetter,
        testSetterWithCallback: testSetterWithCallback,
        testUpdateInt: testUpdateInt,
        testUpdateIntCb: testUpdateIntCb,
    };
}
exports.default = useTypeCheck;
